src/lib/auth.ts

```diff
@@ -275,56 +307,88 @@ export const getClassInfo = (params: GetClassInfoParams) => {
+      const {demoClass, demoUser, demoOffering} = urlParams;
+      if (!demoClass || !demoUser || !demoOffering) {
+        return reject("Missing demoClass or demoUser or demoOffering parameter for demo!");
+      }
+      const [userType, userId, ...rest] = demoUser.split(":");
+      if (((userType !== "student") && (userType !== "teacher")) || !userId) {
+        return reject("demoUser must be in the form of student:<id> or teacher:<id>");
       }
+      resolve(createDemoInfo(demoClass, userType, userId, demoOffering));
+    }
 
-      resolve({authenticatedUser: DEV_USER, classInfo: DEV_CLASS_INFO});
+    if (appMode !== "authed") {
+      resolve({authenticatedUser: DEV_STUDENT, classInfo: DEV_CLASS_INFO});
     }
 
-    if (!token) {
+    if (!bearerToken) {
       return reject("No token provided for authentication (must launch from Portal)");
     }
 
-    if (!domain) {
-      return reject("Missing domain query parameter (required when token parameter is present)");
+    if (urlParams.reportType) {
+      if (urlParams.reportType !== "offering") {
+        return reject("Sorry, only external reports at the offering level are supported");
+      }
+      if (!urlParams.class) {
+        return reject("Missing class parameter!");
+      }
+      if (!urlParams.offering) {
+        return reject("Missing offering parameter!");
+      }
+
+      const {protocol, host} = parseUrl(urlParams.class);
+      basePortalUrl = `${protocol}//${host}/`;
+    }
+    else if (urlParams.domain) {
+      basePortalUrl = urlParams.domain;
+    }
+    else {
+      return reject("Missing domain query parameter!");
     }
 
-    return getPortalJWTWithBearerToken(domain, "Bearer", token)
+    return getPortalJWTWithBearerToken(basePortalUrl, "Bearer", bearerToken)
       .then(([rawPortalJWT, portalJWT]) => {
 
-        return getFirebaseJWTWithBearerToken(domain, "Bearer", token)
+        return getFirebaseJWTWithBearerToken(basePortalUrl, "Bearer", bearerToken)
           .then(([rawFirebaseJWT, firebaseJWT]) => {
 
-            if (portalJWT.user_type === "learner") {
-              const classInfoUrl = portalJWT.class_info_url;
-
-              const domainParser = document.createElement("a");
-              domainParser.href = portalJWT.domain;
-              const portal = domainParser.hostname;
-
-              return getClassInfo({classInfoUrl, rawPortalJWT, portal, offeringId: portalJWT.offering_id})
+            if ((portalJWT.user_type === "learner") || (portalJWT.user_type === "teacher")) {
+              const portal = parseUrl(basePortalUrl).host;
+              let classInfoUrl: string;
+              let offeringId: string;
+
+              if (portalJWT.user_type === "learner") {
+                classInfoUrl = portalJWT.class_info_url;
+                offeringId = `${portalJWT.offering_id}`;
+              }
+              else {
+                // class and offering existence is already validated above
```

Comment:
Is this technically true? It looks like they're only validated if `reportType` exists - which I know it should if a teacher launches from an external activity, but it would be nice to validate the Portal's behavior, if I'm reading this right.

Reply:
Yeah, I was just a little lazy there.  I've refactored it to remove the assumption and added code to ensure the variables have values.